### DISCUSSION TITLE
gstreamer-tests failing to build using demo-image-full?

### DISCUSSION THREAD

**Discussion by tmdarwen** [OE4T/meta-tegra#883] [Q&A] [answered]
I've almost got my project 100% working.  I have a custom C++ application that uses GStreamer and some of the GStreamer "bad" plugins.  Earlier today my custom application built as part of the meta-tegra demo distro image and I ran it on my Jetson but if failed to find some of these "bad" plugins I need, so after various reading on the web I thought I might need a bbappends file specifying these plugins.

So, I made a a bbappend file named gstreamer1.0-plugins-bad_1.1%.bbappend and it simply contains one line: `PACKAGECONFIG += " webrtc webrtcdsp srtp sctp dtls waylandsink"`.  This failed to build though.

So, after some more web reading I thought maybe I need to include the meta-multimedia layer so I added this line to bblayers.conf:  `/home/buildmachine/tegra-demo-distro/layers/meta-multimedia \` and also did the demo-image-full build.

It seems this has gotten me closer to success as I noticed some of my specified plugins being built, but it's failing when trying to build gstreamer-tests:
```
Error:
 Problem: package packagegroup-demo-x11tests-1.0-r0.noarch requires gstreamer-tests, but none of the providers can be installed
  - conflicting requests
  - nothing provides gstreamer1.0-plugins-bad-waylandsink needed by gstreamer-tests-1.0-r0.jetson_xavier_nx_devkit_tx2_nx
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
```

Earlier in the build I see these warnings:
```
WARNING: libpcre2-native-10.34-r0 do_fetch: Failed to fetch URL https://github.com/PhilipHazel/pcre2/releases/download/pcre2-10.34/pcre2-10.34.tar.bz2, attempting MIRRORS if available
WARNING: gstreamer1.0-plugins-bad-1.14.4-r0 do_configure: QA Issue: gstreamer1.0-plugins-bad: invalid PACKAGECONFIG: sctp [invalid-packageconfig]
WARNING: gstreamer1.0-plugins-bad-1.14.4-r0 do_configure: QA Issue: gstreamer1.0-plugins-bad: invalid PACKAGECONFIG: waylandsink [invalid-packageconfig]
WARNING: gstreamer-tests-1.0-r0 do_package_qa: QA Issue: gstreamer-tests rdepends on gstreamer1.0-plugins-bad-waylandsink, but it isn't a build dependency? [build-deps]
```

It seems gstreamer-tests are failing to build to due to the missing "waylandsink" plug-in, however, I'm not sure why the waylandsink package is invalid as the warning says.

Anyone have any ideas?

**Comment by madisongh** [comment]
> So, I made a a bbappend file named gstreamer1.0-plugins-bad_1.1%.bbappend and it simply contains one line: PACKAGECONFIG += " webrtc webrtcdsp srtp sctp dtls waylandsink". This failed to build though.

I think this is your problem, as indicated by the `QA Issue` messages you saw.  It's just `wayland` to get the Wayland-related plugins, not `waylandsink`.

Also note that in the demo distro, by default we use gstreamer 1.14, and that doesn't appear to have SCTP support.  You can modify the distro's configuration file to remove the `include conf/include/gstreamer-1.14.conf` line to switch back to 1.16 if you need to.  Most of the NVIDIA-specific plugins should be compatible with 1.16.

**Comment by tmdarwen** [comment, marked as answer]
Thanks @madisongh as you suggested I updated "layers/meta-tegrademo/conf/distro/tegrademo.conf" and commented out line 36 so it now looks like this:

```
# Gstreamer libraries are passed through to containers when using
# nvidia-docker, so our version of Gstreamer must match the one in
# the stock L4T/JetPack release.
#require conf/include/gstreamer-1.14.conf

SDK_NAME = "${DISTRO}-${TCLIBC}-${SDKMACHINE}-${IMAGE_BASENAME}-${TUNE_PKGARCH}-${MACHINE}"
SDKPATH = "/opt/${DISTRO}/${SDK_VERSION}"
```

I then changed my gstreamer1.0-plugins-bad_1.1%.bbappend to PACKAGECONFIG += " webrtc webrtcdsp srtp sctp dtls wayland". 

However, now opencv is failing to build.  I've pasted the bitbake output below but it's not clear to me what the problem is.

```
|
|
| -- Could NOT find PythonInterp (missing: PYTHON_EXECUTABLE) (Required is at least version "2.7")
| -- Looking for ccache - not found
| -- Found ZLIB: /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/recipe-sysroot/usr/lib/libz.so (found suitable version "1.2.11", minimum required is "1.2.3")
| -- Found ZLIB: /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/recipe-sysroot/usr/lib/libz.so (found version "1.2.11")
| -- Found TBB (env): /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/recipe-sysroot/usr/lib/libtbb.so
| -- CUDA detected: 10.2
| -- Automatic detection of CUDA generation failed. Going to build for all known architectures.
| -- CUDA NVCC target flags: --gpu-architecture=compute_62 --gpu-code=sm_62  --include-path /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/recipe-sysroot/usr/local/cuda-10.2/include --library-path /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/recipe-sysroot/usr/local/cuda-10.2/lib;-gencode;arch=compute_62,code=sm_62;-D_FORCE_INLINES
| -- Could not find OpenBLAS include. Turning OpenBLAS_FOUND off
| -- Could not find OpenBLAS lib. Turning OpenBLAS_FOUND off
| -- Could NOT find Atlas (missing: Atlas_CBLAS_INCLUDE_DIR Atlas_CLAPACK_INCLUDE_DIR Atlas_CBLAS_LIBRARY Atlas_BLAS_LIBRARY Atlas_LAPACK_LIBRARY)
| -- Could NOT find BLAS (missing: BLAS_LIBRARIES)
| -- LAPACK requires BLAS
| -- A library with LAPACK API not found. Please specify library location.
| -- OpenCV Python: during development append to PYTHONPATH: /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/build/python_loader
| -- Caffe:   NO
| -- Protobuf:   NO
| -- Glog:   YES
| -- freetype2:   YES
| -- harfbuzz:    YES
| -- Module opencv_ovis disabled because OGRE3D was not found
| -- No preference for use of exported gflags CMake configuration set, and no hints for include/library directories provided. Defaulting to preferring an installed/exported gflags CMake configuration if available.
| -- Found installed version of gflags: /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/recipe-sysroot/usr/lib/cmake/gflags
| -- Detected gflags version: 2.2.2
| -- Checking SFM deps... TRUE
| -- CERES support is disabled. Ceres Solver for reconstruction API is required.
| -- Module opencv_dnn_objdetect disabled because opencv_dnn dependency can't be resolved!
| abort()ing pseudo client by server request. See https://wiki.yoctoproject.org/wiki/Pseudo_Abort for more details on this.
| Check logfile: /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/pseudo//pseudo.log
| FAILED: build.ninja
| /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/recipe-sysroot-native/usr/bin/cmake -S/home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/git -B/home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/build
| ninja: error: rebuilding 'build.ninja': subcommand failed
| WARNING: /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/temp/run.do_install.3096070:1 exit 1 from 'eval ${DESTDIR:+DESTDIR=${DESTDIR} }VERBOSE=1 cmake --build '/home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/build' "$@" -- ${EXTRA_OECMAKE_BUILD}'
| ERROR: Execution of '/home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/build/tmp/work/armv8a_tegra186-oe4t-linux/opencv/4.1.0-r0/temp/run.do_install.3096070' failed with exit code 1
ERROR: Task (/home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/layers/meta-oe/recipes-support/opencv/opencv_4.1.0.bb:do_install) failed with exit code '1'
NOTE: Tasks Summary: Attempted 6792 tasks of which 6791 didn't need to be rerun and 1 failed.

Summary: 1 task failed:
  /home/buildmachine/NonSys/*********/********/*********Jetson6/tegra-demo-distro/layers/meta-oe/recipes-support/opencv/opencv_4.1.0.bb:do_install
Summary: There was 1 ERROR message shown, returning a non-zero exit code.
```

**Reply by madisongh** [reply]
Yeah, I've run into that before.  The opencv recipe needs a fix to address the pseudo abort problem on rebuilds.  In the meantime, try `bitbake -c clean opencv`, then run your build again, to work around it.

**Reply by tmdarwen** [reply]
Okay, gave that a try and now I'm getting the following error:

```
--> Finished dependency resolution
Error:
 Problem: package packagegroup-demo-x11tests-1.0-r0.noarch requires gstreamer-tests, but none of the providers can be installed
  - conflicting requests
  - nothing provides gstreamer1.0-plugins-bad-waylandsink needed by gstreamer-tests-1.0-r0.jetson_xavier_nx_devkit_tx2_nx
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)

ERROR: Logfile of failure stored in: /home/****/NonSys/****/****/****Jetson6/tegra-demo-distro/build/tmp/work/jetson_xavier_nx_devkit_tx2_nx-oe4t-linux/demo-image-full/1.0-r0/temp/log.do_rootfs.3116112
ERROR: Task (/home/****/NonSys/****/****/****Jetson6/tegra-demo-distro/layers/meta-tegrademo/recipes-demo/images/demo-image-full.bb:do_rootfs) failed with exit code '1'
```

Looks like the gstreamer tests need the wayland plugin but I'm not sure how to get that built and installed.  Any idea?

**Reply by tmdarwen** [reply]
I just realized something: When the build was failing (just before you told me about `bitbake -c clean opencv`) I'd removed wayland from the PACKAGECONFIG list in my gstreamer1.0-plugins-bad_1.1%.bbappend file.

I've added it back in and rebuilt without errors.  However, I still need a few missing GStreamer plugins: opus. vpx, webrtc, dtls, srtp, rptmanager, videotestsrc and audiotestsrc.  These are missing from my distro build, even webrtc is missing despite it being in the PACKAGECONFIG list.

Any idea how I add these missing plugins to the build?

**Reply by madisongh** [reply]
PACKAGECONFIG doesn't control image contents, just how the compilation of a particular recipe is configured. To include one of the generated packages in your image, you need to get it pulled in by your image recipe - either directly by adding the package name to `IMAGE_INSTALL` or one of the other variables that are used for enumerating image contents (e.g., `CORE_IMAGE_EXTRA_INSTALL` for most image recipes), or through runtime dependencies in recipes that need those packages (e.g., the `RDEPENDS_${PN}` settings in the `gstreamer-tests` recipe that you were looking at earlier).

The problem with gstreamer plugins can be knowing what the right package name is.  The plugin packages are usually named `gstreamer1.0-plugins-GROUP-FEATURE`, where `GROUP` is one of `base`, `good`, `bad`, or `ugly`, and `FEATURE` is usually the plugin name that gstreamer uses.  I usually just search around under `$BUILDDIR/tmp/deploy/rpm` to locate them when I need them, after building the gstreamer recipes.

**Reply by tmdarwen** [reply]
Hi @madisongh I'm finding nothing in the `$BUILDDIR/tmp/deploy/rpm` dir that seems to be gstreamer and opus related.  Example:

```
buildmachine@buildmachine:~/tegra-demo-distro$ find build/tmp/deploy/rpm -name *opus*
build/tmp/deploy/rpm/aarch64/libopus-src-1.3.1-r0.aarch64.rpm
build/tmp/deploy/rpm/aarch64/opus-tools-doc-0.1.8-r0.aarch64.rpm
build/tmp/deploy/rpm/aarch64/libopus0-1.3.1-r0.aarch64.rpm
build/tmp/deploy/rpm/aarch64/opus-tools-dev-0.1.8-r0.aarch64.rpm
build/tmp/deploy/rpm/aarch64/libopus-dbg-1.3.1-r0.aarch64.rpm
build/tmp/deploy/rpm/aarch64/opus-tools-dbg-0.1.8-r0.aarch64.rpm
build/tmp/deploy/rpm/aarch64/opus-tools-src-0.1.8-r0.aarch64.rpm
build/tmp/deploy/rpm/aarch64/opus-tools-0.1.8-r0.aarch64.rpm
build/tmp/deploy/rpm/aarch64/libopus-dev-1.3.1-r0.aarch64.rpm
buildmachine@buildmachine:~/tegra-demo-distro$
```

As you mentioned, I would expect to find something more like `gstreamer1.0-plugins-base-opus` but as the list above shows, none of the "opus" files found seem related to gstreamer.

I imagine this is more of a question for the gstreamer group, but you've been very helpful and seem to have a lot of insight into gstreamer and I'm wondering if you have any ideas on this?

**Reply by madisongh** [reply]
A quick search of the gstreamer recipes show 'opus' as a PACKAGECONFIG option in `gstreamer1.0-plugins-base_1.16.3.bb` (in OE-Core dunfell), and `opusparse` as an option for `gstreamer1.0-plugins-bad_1.16.3.bb`.  Neither look to be enabled by default, so you'd need to add them in a bbappend.  Assuming you do that and they build OK, you should find the packages with the plugins you're looking for.

**Reply by tmdarwen** [reply]
Ah, my mistake, I should've found that.  I didn't realize grep didn't search paths that were links and I was grepping the `layers` dir.  I should've been grepping the `repo` dir.  I've added the necessary options to the bb appends files and the plugin is now on the image.  Thank you @madisongh
